### PR TITLE
Feature dataframe construction

### PR DESCRIPTION
**Pull request by lianabagh**

### CODE REVIEW

File: yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py [hunk 1]
```diff
@@ -1,221 +1,348 @@
...
+                           "count",
+                           "relative_mean_harvest",
+                           "std",
+                           "hist"]
+            paths: Dict: raster_type - tile_path
+
+
+        """
+        self.alert_types = ["Low Emergence", "Nutrient Deficiency", "Replant Risk", "New yield risk"]
+        self.low_harvest_threshold = 50
```
Comment (razmikmelikbekyan) [inline yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py:84]: please move above, as a class attribute

File: yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py [hunk 2]
```diff
@@ -1,221 +1,348 @@
...
+        """
+
+        read  all the images and keep it in the dictionary for easy accessing
+
+        """
+        for raster_type in self.paths.keys():
+            if raster_type in frozenset(['planter', 'harvest']):
+                self.tile_rasters[raster_type] = {}
+                for crop, tile_path in self.paths[raster_type].items():
+                    self.tile_rasters[raster_type][crop] = []
```
Comment (razmikmelikbekyan) [inline yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py:147]: why to have list here ? we should have only one planter and harvest type for the given crop type.

File: yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py [hunk 3]
```diff
@@ -1,221 +1,348 @@
...
+
+            if raster_type == 'mask':
+                self.tile_rasters[raster_type] = {}
+                for mask_key, tile_path in self.paths[raster_type].items():
+                    if mask_key == 'boundary':
+                        continue
+                    with rasterio.open(tile_path) as dataset:
+                        self.tile_rasters[raster_type][mask_key] = dataset.read()[0, ...]
+
+            elif raster_type in ['aerial']:  # ('red', 'green', 'blue', 'nir'):
```
Comment (razmikmelikbekyan) [inline yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py:160]: maybe just equal to aerial

File: yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py [hunk 4]
```diff
@@ -1,221 +1,348 @@
...
+
+                self.tile_rasters[raster_type] = []
+                for i, flight in enumerate(self.paths[raster_type]):
+                    self.tile_rasters[raster_type].append({})
+                    for channel_name, tile_path in flight.items():
+                        with rasterio.open(tile_path) as dataset:
```
Comment (razmikmelikbekyan) [inline yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py:166]: make a simple method: `read_raster_tile(raster_path):` that returns the array, by having that here you can have a simple dict comprehension, which will be much faster

File: yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py [hunk 5]
```diff
@@ -1,221 +1,348 @@
...
+                            self.tile_rasters[raster_type][-1][channel_name] = dataset.read()[0, ...]
+
+                    self.tile_rasters['ndvi'] = []
+                    for flight in self.paths['aerial']:
```
Comment (razmikmelikbekyan) [inline yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py:170]: you are doing multiple times looping, you have already flight, no need to loop again

File: yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py [hunk 6]
```diff
@@ -1,221 +1,348 @@
...
+                                  "count": self.add_count,
+                                  "relative_mean_harvest": self.add_means_by_mask}
+
+    def validate_paths(self):
+        """
+
+        Validation of paths which are necessary for processing
+
+        """
+        assert len(set(['aerial', 'planter', 'mask']).intersection(self.paths.keys())) == 3
```
Comment (razmikmelikbekyan) [inline yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py:114]: use is_subset

File: yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py [hunk 7]
```diff
@@ -1,221 +1,348 @@
...
+    with rasterio.open(nir_path) as dataset_nir:
+        nir_tile = dataset_nir.read()[0, ...]
+
+    no_data_mask = ((red_tile < 100) | (nir_tile < 100))
+    ndvi = np.true_divide((nir_tile - red_tile), (nir_tile + red_tile))
+    ndvi[no_data_mask] = -10000
+    return ndvi, dataset_red.meta['nodata']
+
+
+class DataFeatureSample:
```
Comment (razmikmelikbekyan) [inline yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py:55]: what is the main method for getting {crop_type: series}

**Review by razmikmelikbekyan — CHANGES_REQUESTED** [review]

File: yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py (hunk 7 above)
Comment (lianabagh) [reply yield_forecasting/data_preparation/feature_extractors/feature_dataframe_construction.py:55]: add_features

**Review by razmikmelikbekyan — APPROVED** [review]

**Pull request merged**